lsp/Main.hs

```diff
@@ -277,28 +299,42 @@ extract :: Monad m => String -> Maybe a -> ExceptT Err m a
 extract errMessage = except . maybeToRight (Err (DummySRng "From lsp") errMessage)
 
 -- TODO: #65 Show type information on hover
-tokensToHover :: Position -> Program SRng -> ExceptT Err IO Hover
+tokensToHover :: Position -> Program LocAndTp -> ExceptT Err IO Hover
 tokensToHover pos ast = do
       let astNode = findAstAtPoint pos ast
       return $ tokenToHover astNode
 
-tokenToHover :: [SomeAstNode SRng] -> Hover
+tokenToHover :: [SomeAstNode LocAndTp] -> Hover
 tokenToHover astNode = Hover contents range
   where
     astText = astToText astNode
+    dbgInfo2 = tshow $ getChildren $ head astNode
     dbgInfo = case head astNode of
       ast@SProg{} -> T.take 80 $ tshow ast
       ast         -> tshow ast
-    txt = astText <> "\n\n" <> dbgInfo
+    tpInfo = astToTypeInfo astNode
+    txt = astText <> "\n\n" <> tpInfo <> "\n\n" <> dbgInfo <> "\n\n" <> dbgInfo2
     contents = HoverContents $ markedUpContent "haskell" txt
     annRange = getLoc $ head astNode
     range = sRngToRange annRange
 
-astToText :: [SomeAstNode SRng] -> T.Text
+
+astToTypeInfo :: (Show a, HasAnnot f) => [SomeAstNode (f a)] -> T.Text
+astToTypeInfo (SMapping (Mapping tpInfo _ _):_) = tshow $ getAnnot tpInfo
+astToTypeInfo (SClassDecl (ClassDecl tpInfo _ _):_) = tshow $ getAnnot tpInfo
+astToTypeInfo (SGlobalVarDecl (VarDecl tpInfo _ _):_) = tshow $ getAnnot tpInfo
+astToTypeInfo (SRule (Rule tpInfo _ _ _ _):_) = tshow $ getAnnot tpInfo
+astToTypeInfo (SAssertion (Assertion tpInfo _):_) = tshow $ getAnnot tpInfo
+astToTypeInfo ((SExpr e):_) = tshow $ getAnnot $ annotOfExpr e
+astToTypeInfo _ = "No type info found"
```

Comment:
All of these can be reduced to a single line, by using `getAnnot` together with `getType` directly.